src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/vector/view.hpp

```diff
@@ -0,0 +1,281 @@
+    /// Undefined behavior if the vector is empty.
+    reference front() noexcept {
+        return *begin();
+    }
+
+    /// @brief Obtain a const reference to the first element
+    /// @return An element reference
+    /// Undefined behavior if the vector is empty.
+    constexpr const_reference front() const noexcept {
+        return *begin();
+    }
+
+    /// @brief Obtain a reference to the last element
+    /// @return An element reference
+    /// Undefined behavior if the vector is empty.
+    reference back() noexcept {
+        return *(begin() + element_difference_type(size() - 1u));
+    }
+
+    /// @brief Obtain a const reference to the last element
+    /// @return An element reference
+    /// Undefined behavior if the vector is empty.
+    constexpr const_reference back() const noexcept {
+        return *(begin() + element_difference_type(size() - 1u));
+    }
+
+    /// @brief Obtain a byte iterator pointing to the beginning of the vector
+    /// @return A per-byte random access iterator, pointing at the first byte if one exists
+    constexpr byte_iterator byte_begin() const noexcept {
+        return format_traits::make_byte_iterator(begin(), end());
+    }
+
+    /// @brief Obtain a byte iterator pointing just past the end of the vector
+    /// @return A per-byte random access iterator, pointing just past the end of the vector
+    constexpr byte_iterator byte_end() const noexcept {
+        return byte_begin() + byte_difference_type(byte_size());
+    }
+
+    /// @brief Obtain a reference to the first byte
+    /// @return A byte reference
+    /// Undefined behavior if the vector is empty.
+    byte_reference byte_front() noexcept {
+        return *byte_begin();
+    }
+
+    /// @brief Obtain a const reference to the first byte
+    /// @return A byte reference
+    /// Undefined behavior if the vector is empty.
+    constexpr const_byte_reference byte_front() const noexcept {
+        return *byte_begin();
+    }
+
+    /// @brief Obtain a reference to the last byte
+    /// @return A byte reference
+    /// Undefined behavior if the vector is empty.
+    byte_reference byte_back() noexcept {
+        return *(byte_begin() + byte_difference_type(byte_size() - 1u));
+    }
+
+    /// @brief Obtain a const reference to the last byte
+    /// @return A byte reference
+    /// Undefined behavior if the vector is empty.
+    constexpr const_byte_reference byte_back() const noexcept {
+        return *(byte_begin() + byte_difference_type(byte_size() - 1u));
+    }
+
+    /// @brief Obtain a reference to a numbered byte, with bounds checking
+    /// @param index Index in the range 0 to byte_size()-1 inclusive.
+    /// @return A byte reference
+    /// @throws std::out_of_range, if the index is outside the allowed range.
+    byte_reference byte_at(byte_count_type index) {
+        if (index >= byte_size()) {
+            throw std::out_of_range{"bsoncxx::vector::view::byte_at()"};
```

Comment:
This is unusual. All exceptions thrown by the bsoncxx library (with the exception of standard library polyfills) have type `bsoncxx::v_noabi::exception`, [as documented](https://mongocxx.org/api/mongocxx-4.0.0/classbsoncxx_1_1v__noabi_1_1exception.html):

> Class representing any exceptions emitted from the bsoncxx library or its underlying implementation.

Consider `throw bsoncxx::v_noabi::exception{error_code::k_vector_out_of_range};` or similar instead (as is being done with `k_vector_too_large`).